File: internal/ref/lz4.go
package ref

// LZ4 block decoding, the portable reference and the specification for
// simd_lz4_block_decode.
//
// The block format is sequences of
//
//	token | literal-length extension | literals | offset | match-length ext
//
// with four-bit lengths in the token (15 escapes to 255-run extension
// bytes), a two-byte little-endian match offset into the output already
// written, and a minimum match of four. The final sequence carries only
// literals: input ending exactly after them is the one well-formed way a
// block ends.
//
// LZ4BlockDecode returns the decoded length, or -1 for malformed input:
// truncated anywhere, a zero or too-far offset, or output past cap. The
// kernel must agree on the byte and on the -1, and the differential fuzz
// says so.
func LZ4BlockDecode(dst, src []byte) int {
	d, n := 0, len(src)
	i := 0
	for {
		if i >= n {
			return -1 // a block cannot end before a token
		}
		token := src[i]
		i++
		litLen := int(token >> 4)
		if litLen == 15 {
			for {
				if i >= n {
					return -1
				}
				b := src[i]
				i++
				litLen += int(b)
				if b != 255 {
					break
				}
			}
		}
		if litLen > n-i || litLen > len(dst)-d {
			return -1
		}
		copy(dst[d:], src[i:i+litLen])
		d += litLen
		i += litLen
		if i == n {
			return d // the final sequence: literals only
		}
		if n-i < 2 {
			return -1
		}
		offset := int(src[i]) | int(src[i+1])<<8
		i += 2
		if offset == 0 || offset > d {
			return -1
		}
		matchLen := int(token&0xF) + 4
		if token&0xF == 15 {
			for {
				if i >= n {
					return -1
				}
				b := src[i]
				i++
				matchLen += int(b)
				if b != 255 {
					break
				}
			}
		}
		if matchLen > len(dst)-d {
			return -1
		}
		// Overlap is the point: an offset of one replicates a byte. The
		// per-byte loop is the definition; the kernel is what gets clever.
		for k := 0; k < matchLen; k++ {
			dst[d] = dst[d-offset]
			d++
		}
	}
}

File: csrc/lz4.c
// LZ4 block decode: the reference's byte-defined walk with the copies
// widened. Literals move sixteen bytes per store; matches with an offset
// of at least sixteen do too; the short offsets that make LZ4 LZ4 -- an
// offset of one is a run -- replicate per byte, exactly as the reference
// defines them. Every widened store checks its slop against the true
// bounds first and falls back to the byte loop near either end, so the
// kernel writes nothing the reference would not.
//
// Returns the decoded length, or -1 for malformed input: truncated
// anywhere, a zero or too-far offset, output past cap. Agreement with
// the reference -- byte for byte, -1 for -1 -- is the contract the
// differential fuzz enforces.

#include "goabi.h"

typedef long isize;
typedef unsigned char u8;
typedef u8 lzu8x16 __attribute__((ext_vector_type(16), aligned(1)));

void simd_lz4_block_decode(isize *__restrict out, u8 *__restrict dst,
                           isize dcap, const u8 *__restrict src, isize n) {
  isize d = 0, i = 0;
  for (;;) {
    if (i >= n) goto bad;
    u8 token = src[i++];
    isize litLen = token >> 4;
    if (litLen == 15) {
      for (;;) {
        if (i >= n) goto bad;
        u8 b = src[i++];
        litLen += b;
        if (b != 255) break;
      }
    }
    if (litLen > n - i || litLen > dcap - d) goto bad;
    // Wide literal copy when sixteen bytes of slop exist on both sides.
    {
      isize k = 0;
      if (litLen >= 16 && n - (i + litLen) >= 0 && dcap - d >= litLen) {
        for (; k + 16 <= litLen && i + k + 16 <= n && d + k + 16 <= dcap;
             k += 16)
          *(lzu8x16 *)(dst + d + k) = *(const lzu8x16 *)(src + i + k);
      }
      for (; k < litLen; k++) dst[d + k] = src[i + k];
    }
    d += litLen;
    i += litLen;
    if (i == n) {
      *out = d;
      return;
    }
    if (n - i < 2) goto bad;
    isize offset = (isize)src[i] | ((isize)src[i + 1] << 8);
    i += 2;
    if (offset == 0 || offset > d) goto bad;
    isize matchLen = (isize)(token & 0xF) + 4;
    if ((token & 0xF) == 15) {
      for (;;) {
        if (i >= n) goto bad;
        u8 b = src[i++];
        matchLen += b;
        if (b != 255) break;
      }
    }
    if (matchLen > dcap - d) goto bad;
    if (offset >= 16) {
      isize k = 0;
      for (; k + 16 <= matchLen && d + k + 16 <= dcap; k += 16)
        *(lzu8x16 *)(dst + d + k) = *(const lzu8x16 *)(dst + d + k - offset);
      for (; k < matchLen; k++) dst[d + k] = dst[d + k - offset];
    } else {
      for (isize k = 0; k < matchLen; k++) dst[d + k] = dst[d + k - offset];
    }
    d += matchLen;
  }
bad:
  *out = -1;
}
